src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -149,87 +169,158 @@ class FastSync(
-                sys.exit(1)
-                false
-              } else {
-                // validation failed at non-target block, discard N blocks and resume
-                blacklist(peer.id, blacklistDuration, "block header validation failed")
-                discardLastBlocks(header.number, N)
-                syncState = syncState.copy(
-                  blockBodiesQueue = Seq.empty,
-                  receiptsQueue = Seq.empty,
-                  bestBlockHeaderNumber = (header.number - N - 1) max 0,
-                  nextBlockToFullyValidate = (header.number - N) max 1)
-                true // do not process remaining headers and resume
-              }
-          }
+        processHeader(header, peer) match {
+          case Left(result)        => result
+          case Right(headerAndDif) =>
+            updateSyncState(headerAndDif._1, headerAndDif._2)
+            if (header.number == syncState.safeDownloadTarget){
+              ImportedTargetBlock
+            } else {
+              processHeaders(peer, headers.tail)
+            }
+        }
+      } else
+        HeadersProcessingFinished
+    }
+
+    private def validateHeader(header: BlockHeader, peer: Peer): Either[HeaderProcessingResult, BlockHeader] = {
+      val shouldValidate = header.number >= syncState.nextBlockToFullyValidate
+
+      if (shouldValidate) {
+        validators.blockHeaderValidator.validate(header, blockchain) match {
+          case Right(_) =>
+            updateValidationState(header)
+            Right(header)
+
+          case Left(error) =>
+            log.warning(s"Block header validation failed during fast sync at block ${header.number}: $error")
+            Left(ValidationFailed(header, peer))
+        }
+      } else {
+        Right(header)
+      }
+    }
+
+    private def updateSyncState(header: BlockHeader, parentTd: BigInt): Unit = {
+      blockchain.save(header)
+      blockchain.save(header.hash, parentTd + header.difficulty)
+
+      if (header.number > syncState.bestBlockHeaderNumber) {
+        syncState = syncState.copy(bestBlockHeaderNumber = header.number)
+      }
+
+      syncState = syncState
+        .enqueueBlockBodies(Seq(header.hash))
+        .enqueueReceipts(Seq(header.hash))
+    }
+
+    private def updateValidationState(header: BlockHeader): Unit ={
+      import syncConfig.{fastSyncBlockValidationK => K, fastSyncBlockValidationX => X}
+      syncState = syncState.updateNextBlockToValidate(header, K, X)
+    }
+
+    private def processHeader(header: BlockHeader, peer: Peer) : Either[HeaderProcessingResult , (BlockHeader, BigInt)] = for {
+      validatedHeader  <- validateHeader(header, peer)
+      parentDifficulty <- getDifficulty(header)
+    } yield (validatedHeader, parentDifficulty)
+
+    private def getDifficulty(header: BlockHeader) = {
+      blockchain.getTotalDifficultyByHash(header.parentHash).toRight(HeadersProcessingFinished)
```

Comment:
I'm not quite sure I get this. Why does it return `HeadersProcessingFinished` if there's no difficulty for given header?

Reply:
it is just the way to stop processing headers as soon as we encounter one without a parent. 
In earlier code version it `insertHeader` returned `false` and stopped processing restof headers. 
I will make it more transparent what is happening